### DISCUSSION TITLE
LDM Notes for October 16th, 2023

### DISCUSSION THREAD

**Discussion by 333fred** [dotnet/csharplang#7603] [LDM Notes]
https://github.com/dotnet/csharplang/blob/main/meetings/2023/LDM-2023-10-16.md

## Agenda

- Triage
    - Breaking change warnings
    - Determine natural type of method group by looking scope-by-scope
    - u8 string interpolation
    - Lock statement pattern
    - String/Character escape sequence \\e as a short-hand for \\u001b 
    - New operator %% for canonical Modulus operations

**Comment by iam3yal** [comment]
> We think that extension methods are probably the best path forward for getting an infix-style calling pattern for these operations.

Is this being tracked somewhere as part of the new proposal for extensions or something you might consider in the future?

**Reply by CyrusNajmabadi** [reply]
Something we might consider in the future.

**Reply by 333fred** [reply]
This is not something new. This is the existing extension method syntax as it exists today.

**Reply by iam3yal** [reply]
@333fred The wording here suggests that we **might** be getting an alternative syntax at some point or not? to me it reads like Kotlin's infix functions. 😄

**Reply by 333fred** [reply]
No. The wording says that `1.ModFloor(3)` is infix-like.

**Reply by iam3yal** [reply]
@333fred ahh okay, thank you for the clarification. :)

**Comment by DavidArno** [comment]


**Reply by HaloFour** [reply]
I read that to understand that even if combinations of features end up being useless (for some value of useless), such as marking a setter as `readonly`, that it's not necessarily worth it to expend extra effort to forbid it.

**Reply by IS4Code** [reply]
I think the quotes should be always taken a bit less seriously. As far as optional lambda parameters are considered, I definitely have a use for them (exporting functions to an embedded language), though I am a bit confused why the anonymous method syntax (using `delegate`) does not support them for some reason.

In any case, the point is just that if you yourself feel a particular feature is useless, it does not mean it shouldn't be considered. It may be useful to some, so if it doesn't hurt, why not. `readonly` setters seem like something niche but are a perfectly valid logical conclusion to the language rules, so *not doing* them feels like an unnecessary exception.

**Reply by CyrusNajmabadi** [reply]
> what seems to me to be useless features (default values for lambda parameters being the first one that comes to mind)

You're welcome to engage the team on these issues. We're happy to explain why these things are not useless. 

Note: we're not of the opinion that all our features will be useful *to you*. That's a non-goal. But we certainly do believe there is enough value to *the ecosystem as a whole* to warrant it.  If you're looking for an explanation of the former, we likely won't be able to help. But if your mind is open to the latter, we can def provide that.

**Reply by jnm2** [reply]
Readonly setters could be vital to a useful pattern, a wrapper struct around a class instance, where the wrapper struct exposes limited surface area. `CancellationToken` is a common example of a wrapper struct limiting access to a class instance. The wrapper struct itself should be readonly because its contents (the managed address of the class) never change. Imagining wrapping a class with a settable property isn't much of a stretch.

**Reply by CyrusNajmabadi** [reply]
> Imagining wrapping a class with a settable property isn't much of a stretch.

We call these "sinks" when they come up. And they are very reasonable in certain circumstances. :-)

**Reply by CyrusNajmabadi** [reply]
> saves millions of devs

The vast majority of devs don't encounter any sort of 'ref' scenario outside of just a `ref` parameter (and we're already slicing to just a narrow set of devs that even encounter that).  The ref-space is already advanced, self-selected, and intended for users operating in code that needs to understand this stuff to squeeze out all the perf it can.  As such, the users actually working with this are already equipped to learn waht this means (and are normally the users who have been wanting this and asking for it because they need it).

**Reply by HaloFour** [reply]
> The point I was making is that - whether justified or not - there is a feeling amongst some that the language team waste time on useless features and that effectively publicly stating this is the case doesn't help their cause.

That sounds like a problem with those people, and not one that the language team should concern themselves with.  There will always be people who are annoyed that the set of features getting focus aren't their pet set of features.

**Reply by CyrusNajmabadi** [reply]
> there is a feeling amongst some that the language team waste time on useless features

Yes.  This is the norm.  We have no goal to please everyone and make everyone feel that everything we do is teh right set of useful features they want.  To remind you (which has come up numerous times in the past), we have users that have felt that things like 'generics' was useless, or that lambdas were 'useless', etc. etc.   That's *life*.  If you're going to get hung up on our language doing things that you don't always find useful, then you're going to be perpetually peeved.  And that just seems like an issue to resolve outside of this repo.

**Reply by 333fred** [reply]
> there is a feeling amongst some that the language team waste time on useless features and that effectively publicly stating this is the case doesn't help their cause

I'm sorry David, but quotes of the day are amusing anecdotes. If some people can't handle the LDT poking a bit of fun at itself for "useless" features that are actually quite useful and lead to people like Stephen Toub being able to post ".NET is X% faster this year" novellas, I'm not going to concern myself with it.

**Reply by jnm2** [reply]
> This misses the point that I was trying to make. I've no idea what a readonly setter is supposed to be and I don't really care either. It sounds silly to me, but it could well be very useful. The point I was making is that - whether justified or not - there is a feeling amongst some that the language team waste time on useless features and that effectively publicly stating this is the case doesn't help their cause. That's all.

@DavidArno The measure of the value of my reply is not whether it connects with the point you had in mind. It would be presumptive for you to assume so. Rather, I'm rejecting the premise. I think it's useful for people to take that mental path of questioning such premises, and so I shared this use case.

I think your reaction to the quote points to something interesting, but I really believe there's not much for me to say about it. I agree with the others here; the angst looks self-imposed, and the cited harm of millions of devs noticing that `readonly set` exists and losing sleep over it or over any other language curiosity, does not sound convincing.

**Reply by HaloFour** [reply]
I don't see it as any more unprofessional than any of the tongue-in-cheek quotes that the team has included in their notes.  I'd rather the team be capable of poking fun at themselves than to be devoid of humor and humanity.  Rather it seems that you're obsessed with this notion that this particular quote is an admission that the team recognizes that the features that they work on is useless, which is why you attributed it towards other features you disagree with but that the team didn't mention at all?

**Reply by CyrusNajmabadi** [reply]
@DavidArno Honestly, you might want to consider just having a blog or something.

**Reply by iam3yal** [reply]
@DavidArno You really, truly believe they are paid to waste time? even rarely? I feel like you're being a bit unfair also aren't you reading too much into these quotes? I mean they are allowed to have fun. 😄

**Reply by CyrusNajmabadi** [reply]
> and that effectively publicly stating this is the case doesn't help their cause.

We're not stating anything publicly. It was a quote of a joke told between colleagues who have a healthy working relationship and lots of camaraderie.  The notes include quips and jokes because they are fun, and because literally everyone else recognizes and can discern between that and the actual business at hand.  

I'll repeat: if this is a problem for you, then that's not something we can help with. Feel free to vent on a blog or something.

**Reply by iam3yal** [reply]
@DavidArno Alright. 🤦‍♂️  Peace. Have a nice day. 😄

**Reply by spydacarnage** [reply]
@DavidArno 

As your extension blocks everything posted by Cyrus, I'll repeat his comment for you...

> and that effectively publicly stating this is the case doesn't help their cause.

We're not stating anything publicly. It was a quote of a joke told between colleagues who have a healthy working relationship and lots of camaraderie. The notes include quips and jokes because they are fun, and because literally everyone else recognizes and can discern between that and the actual business at hand.

I'll repeat: if this is a problem for you, then that's not something we can help with. Feel free to vent on a blog or something.

**Reply by jnm2** [reply]
That's an incredibly mean-spirited thing to say. David goes way past crossing a line of decency in my opinion.

**Reply by En3Tho** [reply]
@CyrusNajmabadi They say the sky is the limit. But if we're talking about patience in particular then it's most probably you :D

**Reply by amis92** [reply]
@DavidArno if your protein-based CPU can't handle skipping/ignoring some messages (that are relevant for full discussion context) and you need silicone-based CPU to help with that, maybe also use silicone CPU to perform transformations of your own protein-CPU outputs with some "positiveness-enhancing" filter, and even let it fully skip posting of some of those outputs.

**Reply by jnm2** [reply]
> Do you genuinely think I should have kept quiet about it and left Cyrus potentially wasting his time writing things that won't be read by the intended audience?
>
> Even when I try to be helpful, I still get attacked. 😞

@DavidArno I believe you're bringing down the level of discourse in this repo through inflammatory remarks while believing yourself to be under attack. My heart sinks when I see it. You're seeing enemies where there are none, and you're not being fair to Cyrus, me, or others, and really none of this is about C#. From what I and others can tell, it's about you.

**Reply by AndersKehlet** [reply]
I, for one, enjoy the quotes and hope this bit of drama won't affect future decisions in that regard.

**Comment by TheUnlocked** [comment]
The modulus decision is a bit disappointing. While I agree that there are any number of potential boutique operators people might want following something like this, canonical modulus has particular practical value that other potential operators in the area don't.

BTW, should this decision also be taken as a death knell for #2585?

**Reply by CyrusNajmabadi** [reply]
One man's "practical value" is another's "boutique operator" :-)

> BTW, should this decision also be taken as a death knell for

No. No decision was made on that.

**Reply by sab39** [reply]
I agree with OP, I haven't felt so strongly that an LDM decision is flat _wrong_ in a long time (possibly since the plan to release `Nullable<T>` without runtime support in 2005, although obviously this one is far lower stakes).

I understand that the importance of any given feature is subjective, but I can think of dozens of times I've wanted to use canonical modulus in general-purpose non-specialized algorithms and code, and I can't see any applications at all for what we have today outside of actual mathematics.

```cs
var bucketIndex = item.GetHashCode() % buckets.Count;
DayOfWeek AddDays(this DayOfWeek day, int days) => (DayOfWeek) (((int) day + days) % 7);
```

If not for having followed this discussion - and for having run into bugs caused by the current behavior - it would never occur to me that `0 <= a % b < b` wasn't an invariant for any positive `b`. And I definitely am far more obsessively interested in precise details of programming language semantics than most developers!

**Reply by CyrusNajmabadi** [reply]
> but I can think of dozens of times I've wanted to use canonical modulus

Right.  Any decision we make to not do something will likely not be viewed positively by a particular customer that woudl have made lots of use of that feature.  So we have to look more broadly, otherwise we literally would have to do *all* features.

**Reply by sab39** [reply]
> Right. Any decision we make to not do something will likely not be viewed positively by a particular customer that woudl have made lots of use of that feature. So we have to look more broadly, otherwise we literally would have to do _all_ features.

Honestly, I'd be mollified if I saw someone make a plausible claim that using `%` for anything _other_ than wrapping a value into the range [0, n) accounted for even a significant _minority_ of real world code, let alone a majority. I know that my experience is always going to be limited and unrepresentative, so it's entirely possible I _am_ unaware of a widespread need for a remainder operator in a type of code I'm not familiar with, and if so then the decision becomes a lot more subjective, even if I still disagree with it.

One other point: considering the widespread confusion about the difference between the two operations, having a separate operator would have a kind of educational or documentation-like value, in that, regardless of which of the operators a programmer actually needs for their algorithm, they would have the chance to choose it intentionally rather than potentially guessing wrong about what the operator does.

**Reply by tannergooding** [reply]
> I haven't felt so strongly that an LDM decision is flat wrong in a long time

The decision to not expose `remainder of floored division` as a language operator doesn't negate the ability for the BCL to provide library functions that support the same and that is why https://github.com/dotnet/runtime/issues/93568 was correspondingly opened.

This is, by far, not unique to C#/.NET and is something that almost every language deals with in some capacity and that has opted to deal with it in a similar manner to how LDM decided here. That is, languages do not agree on whether division and the corresponding remainder operation should be `truncating`, `floored`, or `Euclidean`. There is even at least one that provide the `ceiling` based alternative.

Which of them is correct entirely depends on your payload. Most languages do not provide multiple variants and leave it up to the library to provide alternatives to their default. Of those that do provide both, it's normally differentiated as `x rem y` vs `x mod y` and not done via symbols (i.e. `x % y` and `x %% y`).

>  I can't see any applications at all for what we have today outside of actual mathematics.

Computer math is ultimately not the same as real math. Computers are operating with many more constraints and as such we have to consider scenarios like rounding and overflow as a part of every operation. We do not have access to infinite precision, unbounded range, or even the ability to trivial replace concepts with abstract symbols. Instead, we work with finite precision, finite range, finite time, finite memory/resources, etc.

Because of this, computers have caused an introduction and consideration of all kinds of new or additional rules that tweak the invariants you may otherwise be used to. You have to consider that `x + 1 > x` may not hold true. You have to consider that `x + 1` may be significantly more expensive than `y + 1`. You have to consider that `(a + b) + c` and `a + (b + c)` may not be equivalent. You have to consider that every operation may introduce additional error, and so on.

Correspondingly, this also means thinking about ways you can achieve the same functionality differently or more efficiently. Hardware opted for truncated division because it is simply cheaper than any other kind of division. Remainder of truncated division then naturally falls out of that using a consistent algorithm as for other types of remainders. You can then use these "cheapest foundational" operations to resolve the other types of division or remainders for only a minor additional cost.
